### PR TITLE
Use new container for global layout image map.

### PR DESCRIPTION
**Pull request by locke-lunarg**

### CODE REVIEW

File: layers/core_validation.h [hunk 1]
```diff
@@ -35,9 +35,7 @@ class CoreChecks : public ValidationStateTracker {
     std::unordered_set<uint64_t> ahb_ext_formats_set;
     GlobalQFOTransferBarrierMap<VkImageMemoryBarrier> qfo_release_image_barrier_map;
     GlobalQFOTransferBarrierMap<VkBufferMemoryBarrier> qfo_release_buffer_barrier_map;
-    unordered_map<VkImage, std::vector<ImageSubresourcePair>> imageSubresourceMap;
-    using ImageSubresPairLayoutMap = std::unordered_map<ImageSubresourcePair, IMAGE_LAYOUT_STATE>;
-    ImageSubresPairLayoutMap imageLayoutMap;
+    GlobalImageLayoutMap imageLayoutMap;
```
Comment (locke-lunarg) [inline layers/core_validation.h]: Changed to the new container.
```
typedef ImageSubresourceLayoutMap::LayoutMap GlobalImageLayoutRangeMap;
typedef std::unordered_map<VkImage, std::unique_ptr<GlobalImageLayoutRangeMap>> GlobalImageLayoutMap;
```

File: layers/buffer_validation.cpp [hunk 2]
```diff
@@ -213,151 +213,46 @@ VkImageSubresourceRange NormalizeSubresourceRange(const IMAGE_STATE &image_state
...
-        return false;
-    }
-    imgpair.subresource.aspectMask = aspectMask;
-    auto imgsubIt = imageLayoutMap.find(imgpair);
-    if (imgsubIt == imageLayoutMap.end()) {
+    const auto *layout_range_map = GetLayoutRangeMap(imageLayoutMap, image);
+    if (!layout_range_map) return false;
+    // TODO: what is this test and what is it supposed to do?! -- the logic doesn't match the comment below?!
+
+    // TODO: Make this robust for >1 aspect mask. Now it will just say ignore potential errors in this case.
```
[the remark this answers is not in the document]
Comment (jzulauf-lunarg) [reply layers/buffer_validation.cpp]: This should be fixed in a follow up PR.

File: layers/buffer_validation.cpp [hunk 3]
```diff
@@ -3367,55 +3251,105 @@ void CoreChecks::PreCallRecordCmdBlitImage(VkCommandBuffer commandBuffer, VkImag
...
+        sparse_container::parallel_iterator<const ImageSubresourceLayoutMap::LayoutMap> current_layout(*overlay_map, *global_map,
+                                                                                                       pos->first.begin);
+        while (pos != end) {
+            VkImageLayout initial_layout = pos->second;  // Point directly into the RangeMap iterated but the view
+            VkImageLayout image_layout = kInvalidLayout;
+            // current_layout.seek(pos->first.begin);
+            if (current_layout->range.empty()) break;  // When we are past the end of data in overlay and global... stop looking
+            if (current_layout->pos_A->valid) {
+                image_layout = current_layout->pos_A->lower_bound->second;
+            } else if (current_layout->pos_B->valid) {
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/buffer_validation.cpp]: pos_B?

[the remark this answers is not in the document]
Comment (jzulauf-lunarg) [reply layers/buffer_validation.cpp]: Yes `pos_B` (cut and paste error...)

File: layers/buffer_validation.cpp [hunk 4]
```diff
@@ -213,151 +213,46 @@ VkImageSubresourceRange NormalizeSubresourceRange(const IMAGE_STATE &image_state
 }
 
 template <class OBJECT, class LAYOUT>
-void CoreChecks::SetLayout(OBJECT *pObject, VkImage image, VkImageSubresource range, const LAYOUT &layout) {
-    ImageSubresourcePair imgpair = {image, true, range};
-    SetLayout(pObject, imgpair, layout, VK_IMAGE_ASPECT_COLOR_BIT);
-    SetLayout(pObject, imgpair, layout, VK_IMAGE_ASPECT_DEPTH_BIT);
-    SetLayout(pObject, imgpair, layout, VK_IMAGE_ASPECT_STENCIL_BIT);
-    SetLayout(pObject, imgpair, layout, VK_IMAGE_ASPECT_METADATA_BIT);
+void CoreChecks::SetLayout(OBJECT *pObject, VkImage image, const VkImageSubresource &subresource, const LAYOUT &layout) {
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/buffer_validation.cpp]: fixed it.

File: layers/buffer_validation.cpp [hunk 5]
```diff
@@ -3367,55 +3251,105 @@ void CoreChecks::PreCallRecordCmdBlitImage(VkCommandBuffer commandBuffer, VkImag
...
+
 // This validates that the initial layout specified in the command buffer for the IMAGE is the same as the global IMAGE layout
-bool CoreChecks::ValidateCmdBufImageLayouts(const CMD_BUFFER_STATE *pCB, const ImageSubresPairLayoutMap &globalImageLayoutMap,
-                                            ImageSubresPairLayoutMap *overlayLayoutMap_arg) const {
+bool CoreChecks::ValidateCmdBufImageLayouts(const CMD_BUFFER_STATE *pCB, const GlobalImageLayoutMap &globalImageLayoutMap,
+                                            GlobalImageLayoutMap *overlayLayoutMap_arg) const {
     if (disabled.image_layout_validation) return false;
     bool skip = false;
-    ImageSubresPairLayoutMap &overlayLayoutMap = *overlayLayoutMap_arg;
+    GlobalImageLayoutMap &overlayLayoutMap = *overlayLayoutMap_arg;
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/buffer_validation.cpp]: fixed it.

File: layers/buffer_validation.cpp [hunk 6]
```diff
@@ -3367,55 +3251,105 @@ void CoreChecks::PreCallRecordCmdBlitImage(VkCommandBuffer commandBuffer, VkImag
...
+        // if (global_map->empty() && overlay_map->empty()) // skip this next loop...;
+
+        auto pos = initial_layout_map.begin();
+        const auto end = initial_layout_map.end();
+        sparse_container::parallel_iterator<const ImageSubresourceLayoutMap::LayoutMap> current_layout(*overlay_map, *global_map,
+                                                                                                       pos->first.begin);
+        while (pos != end) {
+            VkImageLayout initial_layout = pos->second;  // Point directly into the RangeMap iterated but the view
+            VkImageLayout image_layout = kInvalidLayout;
+            // current_layout.seek(pos->first.begin);
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/buffer_validation.cpp]: fixed it.

File: layers/buffer_validation.cpp [hunk 7]
```diff
@@ -3367,55 +3251,105 @@ void CoreChecks::PreCallRecordCmdBlitImage(VkCommandBuffer commandBuffer, VkImag
...
+
+        // Note: don't know if it would matter
+        // if (global_map->empty() && overlay_map->empty()) // skip this next loop...;
+
+        auto pos = initial_layout_map.begin();
+        const auto end = initial_layout_map.end();
+        sparse_container::parallel_iterator<const ImageSubresourceLayoutMap::LayoutMap> current_layout(*overlay_map, *global_map,
+                                                                                                       pos->first.begin);
+        while (pos != end) {
+            VkImageLayout initial_layout = pos->second;  // Point directly into the RangeMap iterated but the view
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/buffer_validation.cpp]: fixed it.

File: layers/buffer_validation.cpp [hunk 8]
```diff
@@ -3367,55 +3251,105 @@ void CoreChecks::PreCallRecordCmdBlitImage(VkCommandBuffer commandBuffer, VkImag
...
+            VkImageLayout image_layout = kInvalidLayout;
+            // current_layout.seek(pos->first.begin);
+            if (current_layout->range.empty()) break;  // When we are past the end of data in overlay and global... stop looking
+            if (current_layout->pos_A->valid) {
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/buffer_validation.cpp]: fixed it.

File: layers/buffer_validation.cpp [hunk 9]
```diff
@@ -3367,55 +3251,105 @@ void CoreChecks::PreCallRecordCmdBlitImage(VkCommandBuffer commandBuffer, VkImag
...
         }
 
         // Update all layout set operations (which will be a subset of the initial_layouts
-        for (const auto &pos : subres_map->CurrentLayoutView()) {
-            VkImageLayout layout = pos.it->second;
-            assert(layout != kInvalidLayout);
-            isr_pair.subresource = pos.subresource;
-            SetLayout(overlayLayoutMap, isr_pair, layout);
-        }
+        sparse_container::splice(overlay_map, subres_map->GetCurrentLayoutMap(), sparse_container::value_precedence::prefer_source);
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/buffer_validation.cpp]: fixed it.

File: layers/core_validation.cpp [hunk 10]
```diff
@@ -2158,13 +2165,12 @@ bool CoreChecks::ValidateMaxTimelineSemaphoreValueDifference(VkQueue queue, VkSe
 }
 
 bool CoreChecks::ValidateCommandBuffersForSubmit(VkQueue queue, const VkSubmitInfo *submit,
-                                                 ImageSubresPairLayoutMap *localImageLayoutMap_arg,
-                                                 QueryMap *local_query_to_state_map,
+                                                 GlobalImageLayoutMap *localImageLayoutMap_arg, QueryMap *local_query_to_state_map,
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/core_validation.cpp]: fixed it.

File: layers/buffer_validation.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/buffer_validation.cpp]: fixed it.

File: layers/buffer_validation.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/buffer_validation.cpp]: Added a comment.

**Pull request merged**